visualizer/semanticsHelper.js

```diff
@@ -259,12 +287,81 @@
     return pexpr;
   }
 
+  // Build the action that called from the wrapper, which includes renaming the
+  // operation arguments, and actual action body that user typed.
+  function buildAction(opArgs, enclosedActionArgStr, actionBody) {
+    // Constructs the codes that rename the operation arguments, so we could refer
+    // them by name directly. i.e. `var <argument name> = this.args.<argument name>`
+    var renameOpArgStr = opArgs &&  Object.keys(opArgs).map(function(argName) {
+      return 'var ' + argName + ' = this.args.' + argName + ';';
+    }).join('\n');
+
+    var actionStr = 'function' + enclosedActionArgStr + '{\n' +
+        renameOpArgStr + '\n' + actionBody +
+      '\n}';
+    return eval('(' + actionStr + ')');    // eslint-disable-line no-eval
+  }
+
+  function wrapAction(currentOpName, enclosedActionArgStr, actionBody) {
+    if (!actionBody.trim()) {
+      return undefined;
+    }
+
+    var realAction = 'function' + enclosedActionArgStr + '{\n' + actionBody + '\n}';
+    var isExpression = es6.match(actionBody, 'AssignmentExpression<withIn>').succeeded();
+    if (isExpression) {
+      actionBody = 'return ' + actionBody + ';';
+    }
+    var wrapper = function() {
+      var key = nodeKey(this);
+      var result;
+      var origTodoList = todoList;
+      todoList = undefined;
+      try {
+        var action = buildAction(this.args, enclosedActionArgStr, actionBody);
+        result = action.apply(this, arguments);
+        if (todoList) {
+          result = failure;
+        }
+      } catch (error) {
+        if (todoList) {
+          result = failure;
+        } else {
+          result = error.isErrorWrapper ? error : new ErrorWrapper(key, error);
```

Comment:
Ok, this control flow is *really* tricky to understand. In generally you should use `finally` with caution, because it's really hard to understand to follow the flow of code like this. Can you rewrite it in a way that doesn't use a finally block? I'll leave a few more comments with suggestions.